kroxylicious-kms-provider-aws-kms/src/test/java/io/kroxylicious/kms/provider/aws/kms/credentials/Ec2MetadataCredentialsProviderTest.java

```diff
@@ -0,0 +1,320 @@
+                                .withBody(toJson(initial))));
+
+        try (var provider = new Ec2MetadataCredentialsProvider(config, executorService, Clock.systemUTC())) {
+            var credentialStage = provider.getCredentials();
+            assertThat(credentialStage)
+                    .succeedsWithin(Duration.ofSeconds(1))
+                    .returns("initialKey", SecurityCredentials::secretAccessKey)
+                    .isNotNull();
+
+            var refreshed = createTestCredential("Success", "accessKeyId", "refreshedKey", "token", now.plusSeconds(20));
+            metadataServer.stubFor(
+                    get(urlEqualTo(META_DATA_IAM_SECURITY_CREDENTIALS_ENDPOINT + IAM_ROLE))
+                            .willReturn(WireMock.aResponse()
+                                    .withBody(toJson(refreshed))));
+
+            await().atMost(Duration.ofSeconds(5))
+                    .untilAsserted(() -> {
+                        var refreshedStage = provider.getCredentials();
+                        assertThat(refreshedStage)
+                                .succeedsWithin(Duration.ofSeconds(1))
+                                .returns("refreshedKey", SecurityCredentials::secretAccessKey);
+                    });
+        }
+    }
+
+    @Test
+    void expiredCredentialRefreshed() {
+        var clock = mock(Clock.class);
+        var executor = mock(ScheduledThreadPoolExecutor.class);
+
+        var now = Instant.now();
+        var initial = createTestCredential("Success", "accessKeyId", "initialKey", "token", now.plusSeconds(10));
+
+        doAnswer(invocation -> {
+            var runnable = ((Runnable) invocation.getArguments()[0]);
+            runnable.run();
+            return null;
+        }).when(executor).execute(any(Runnable.class));
+
+        metadataServer.stubFor(
+                get(urlEqualTo(META_DATA_IAM_SECURITY_CREDENTIALS_ENDPOINT + IAM_ROLE))
+                        .willReturn(WireMock.aResponse()
+                                .withBody(toJson(initial))));
+
+        try (var provider = new Ec2MetadataCredentialsProvider(config, executor, clock)) {
+            var credentialStage = provider.getCredentials();
+            assertThat(credentialStage)
+                    .succeedsWithin(Duration.ofSeconds(1))
+                    .returns("initialKey", SecurityCredentials::secretAccessKey)
+                    .isNotNull();
+
+            var refreshed = createTestCredential("Success", "accessKeyId", "refreshedKey", "token", now.plusSeconds(20));
+            metadataServer.stubFor(
+                    get(urlEqualTo(META_DATA_IAM_SECURITY_CREDENTIALS_ENDPOINT + IAM_ROLE))
+                            .willReturn(WireMock.aResponse()
+                                    .withBody(toJson(refreshed))));
+
+            var timeBeyondInitialExpiry = initial.expiration().plusSeconds(1);
+            when(clock.instant()).thenReturn(timeBeyondInitialExpiry);
```

Comment:
Does this do what we intend? I'm not clear at what point `when` conditions are evaluated. I thought they were global but not turning up any references for that.

Might be clearer to have ` when(clock.instant()).thenReturn(now, initial, timeBeyondInitialExpiry);` up around line 196.

Reply:
> Does this do what we intend? I'm not clear at what point when conditions are evaluated. I thought they were global but not turning up any references for that.

It is legal to override Mockito stubbing.  The javadoc for `org.mockito.Mockito#whe`n says:

> Stubbing can be overridden (snip) Last stubbing is more important - when you stubbed the same method with the same arguments many times.

I accept that you could setup the returns in a oner in advance, but for me the test code would be less obvious.  I like the fact that I see the test move the hands on clock where it does.

Reply:
>Stubbing can be overridden (snip) Last stubbing is more important - when you stubbed the same method with the same arguments many times.

Yes last write wins, but when is that write evaluated? Which is what I was getting at. I think your model is probably right and is probably more readable.

Reply:
Ah.. well the implementation reads the clock every time getCredential() is called.  It makes sure that the credential it is about to return isn't expired already.